File: cpp/src/lut_runtime.cpp
#include "utils.cpp"
#include <benchmark/benchmark.h>
#include <chrono>
#include <cmath>
#include <fstream>

int main() {

  std::ofstream perf_ctl("/tmp/perf.ctl");
  std::ifstream perf_ack("/tmp/perf.ack");

  const auto test_cases =
      utils::read_array_from_file<utils::TEST_SIZE>("lookup.txt");

  std::print(
      "LUT size: {}, increment: {}, testSize: {}, degrees: {}, steps: {}\n",
      "N/A", utils::INCREMENT, utils::TEST_SIZE, utils::DEGREES, utils::STEPS);

  // 1 ------ WARMUP LOOP ------
  double warmup_sum = 0.0;
  double sum = 0.0;
  size_t ITERS = 100;

  for (auto i = 0; i < ITERS; i++)
    for (const auto &num : test_cases) {
      double float_num = static_cast<double>(num);
      warmup_sum += std::sin(float_num) + std::cos(float_num);
    }

  benchmark::DoNotOptimize(warmup_sum);

  // 2 ---- BENCHMARK LOOP -----

  // start perf, then the clock
  utils::send_perf_cmd(perf_ctl, perf_ack, "enable");
  auto start_time = std::chrono::steady_clock::now();

  for (auto i = 0; i < ITERS; i++)
    for (const auto &num : test_cases) {
      double float_num = static_cast<double>(num);
      sum += std::sin(float_num) + std::cos(float_num);
    }

  benchmark::DoNotOptimize(sum);

  // end the clock, then stop perf
  auto end_time = std::chrono::steady_clock::now();
  utils::send_perf_cmd(perf_ctl, perf_ack, "disable");

  const auto duration = (end_time - start_time).count();
  std::print("Processed in: [{}] ns. Sum: {}, Warmup Sum: {}\n", duration, sum,
             warmup_sum);

  return 0;
}

File: rust/src/lut_runtime.rs
use std::fs::OpenOptions;
use std::time::Instant;
mod utils;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut perf_ctl = OpenOptions::new().write(true).open("/tmp/perf.ctl")?;
    let mut perf_ack = OpenOptions::new().read(true).open("/tmp/perf.ack")?;

    let test_cases: [f64; utils::TEST_SIZE] =
        utils::read_array_from_file::<{ utils::TEST_SIZE }>("lookup.txt");

    println!(
        "increment: {}, testSize: {}, degrees: {}, steps: {}\n",
        utils::INCREMENT,
        utils::TEST_SIZE,
        utils::DEGREES,
        utils::STEPS,
    );

    let mut warmup_sum: f64 = 0.0;
    let mut sum: f64 = 0.0;

    const ITERS: usize = 100;

    for _ in 0..ITERS {
        for &num in test_cases.iter() {
            warmup_sum += num.cos() + num.sin();
        }
    }

    // finish this calculation
    std::hint::black_box(warmup_sum);

    utils::send_perf_cmd(&mut perf_ctl, &mut perf_ack, "enable")?;
    let start_time = Instant::now();

    for _ in 0..ITERS {
        for &num in test_cases.iter() {
            sum += num.cos() + num.sin();
        }
    }

    std::hint::black_box(sum);

    let end_time = Instant::now();
    utils::send_perf_cmd(&mut perf_ctl, &mut perf_ack, "disable")?;

    let duration = end_time.duration_since(start_time).as_nanos();

    println!(
        "Processed in: [{}] ns, sum: {}, warmup sum: {}",
        duration, sum, warmup_sum
    );

    Ok(())
}
